Automatic gridware updates (closes #121)

bobwhitelock opened the pull request:
Along with a couple of small bug fixes, this PR adds automatic updates after a configurable period of time to gridware, as discussed in https://github.com/alces-software/clusterware/issues/121.

As part of developing this I've set up some testing of gridware and written some tests of this new functionality; I found this helped with developing this since there are numerous different combinations of things that can occur and effect things (e.g. different actions/ repository states/ results of syncing repositories etc.), some of which are also tricky to reliably recreate in a real environment. Tests can be run with `make -f /media/host/clusterware/Makefile` within a standard Clusterware development VM (I've used make rather than just running the rake task directly so that if we decide to automate testing of any other parts of Clusterware we can just slot this in here).

Also, I've added a couple of new values to the `gridware.yml` config file for configuring this feature; these are:
- `:update_period` - how often in days gridware will automatically update repositories when an action requiring an update is run;
- `:last_update_filename` - the filename which will be used within each repository to store the date/time of the last update of this repository.

Do I need to update anywhere to include these in the generated config file for new Clusterware installations? These may need documenting for users somewhere also.

bobwhitelock committed:
Setup gridware testing

So I can attempt not to break things.

Currently run by running
`make -f /media/host/clusterware/Makefile` as `root` in a Clusterware
VM.

bobwhitelock committed:
Add tests for existing depot shortcuts behaviour

bobwhitelock committed:
Fix error for `alces gridware depot list[` etc

bobwhitelock committed:
Refactor method

bobwhitelock committed:
Set up ruby testing so runs all tests

Just use built-in rake task for this since it's already there.

bobwhitelock committed:
Add reading update period in seconds to config

bobwhitelock committed:
Factor out setting up environment to Rake task

Importing `alces/packager/cli` still seems to be required in test files
to avoid import errors for some reason.

bobwhitelock committed:
Add reading last update time to handler proxy

bobwhitelock committed:
Test HandlerProxy calls Handler method

bobwhitelock committed:
Raise an error if time for an update (for now)

Also switch to using DateTime over Time so can just use addition to get
days since. Ruby having 3 date/time APIs is confusing.

bobwhitelock committed:
Allow running Ruby tests from anywhere

By shifting to ruby directory to run.

bobwhitelock committed:
Wrap configured last_update_file in a File

bobwhitelock committed:
Update all repos when time for an update

bobwhitelock committed:
Improve HandlerProxy tests

bobwhitelock committed:
Only perform update before certain actions

These are the actions that it has been decided require a recent update before
they should execute, for other actions this doesn't need to happen.

bobwhitelock committed:
Extract method to handle actions

bobwhitelock committed:
Remove unneeded parameters

bobwhitelock committed:
Always stub update_all in tests

To prevent accidentally actually calling if called unexpectedly.

bobwhitelock committed:
Add methods for having per-repo last_update file

bobwhitelock committed:
Add method to get repos requiring an update

bobwhitelock committed:
Factor mock repos into own module

bobwhitelock committed:
Only update repos requiring an update

Each repository now has its own file tracking the last time it was
updated, as each can be updated independently, so now when an action
requires repositories be up-to-date we only update those repositories
which have not recently been updated.

bobwhitelock committed:
Improve terminology for mock repos

bobwhitelock committed:
Return earliest datetime when no last update

bobwhitelock committed:
Correctly set variable for sync response

Presumably this was set in the past but has since been unset, but this
is needed otherwise a runtime error due to `r` being unset in the
`r.chomp` below occurs when the sync gives an unrecognized response.

bobwhitelock committed:
Set last update time when updating

I have assumed for the moment we want to set this for all situations
when updating except if we get an unexpected response from the git sync
(e.g. if this fails), however this may change in future.

bobwhitelock committed:
Print message when auto-updating; also refactoring

bobwhitelock committed:
Remove no longer needed functionality

This has all been made unnecessary by later changes.

mjtko commented on lib/ruby/lib/alces/packager/config.rb:
```diff
@@ -54,6 +54,10 @@ def dbroot(depot)
         end
 
+        def last_update_file
+          File.new(method_missing(:last_update_file))
```
I think this might be better as `config[:last_update_file]` rather than going via `method_missing`.

bobwhitelock replied:
This would be better, although since you've pointed this out I've realised this method isn't actually being used any more as each repo now has its own last update file (see https://github.com/alces-software/clusterware/blob/f82db8c553ed84fea853608dc9aa800993640694/lib/ruby/lib/alces/packager/repository.rb#L226,L228). I'll remove this method now, thanks for pointing it out.

mjtko commented on lib/ruby/lib/alces/packager/repository.rb:
```diff
@@ -125,12 +131,14 @@ def packages
       def update!
         if metadata.key?(:source)
-          case Alces.git.sync(repo_path, metadata[:source])
+          case r = Alces.git.sync(repo_path, metadata[:source])
```
This assignment doesn't seem to be used.

bobwhitelock replied:
It's used [here](https://github.com/alces-software/clusterware/pull/168/files#diff-8fc2d801996e2afb524f8fffd7c86717R156); I've not changed this line but it was giving an error about `r` being unset (rather than the unrecognized response), so I assumed `r` was supposed to be set to the response here.

mjtko replied:
Ah, good spot! Ta. :)

bobwhitelock committed:
Group test gems

mjtko commented:
As the test gems have now been grouped, I think it would be worthwhile updating the build section of `scripts/dependencies/components.sh` to `--exclude test` on the `bundle install` line (or whatever the syntax actually is :smile:).

mjtko commented:
Also, can you update and relocate the `Makefile` into `scripts` please?  The `scripts` directory contains auxiliary files and makes it easier to keep things a bit tidier when installating.

bobwhitelock committed:
Move Makefile to scripts dir

bobwhitelock committed:
Do not install test gems with dependencies

bobwhitelock commented:
I've made those changes, though in a fresh VM now (after installing the test dependencies) running the tests gives the following error:

```
[root@login1(el7cluster) ruby]# /opt/clusterware/opt/ruby/bin/rake test --trace
** Invoke test (first_time)
** Execute test
/opt/clusterware/lib/ruby/vendor/ruby/2.2.0/gems/json-1.8.3/tests/test_json.rb:4:in `require': cannot load such file -- test/unit (LoadError)
        from /opt/clusterware/lib/ruby/vendor/ruby/2.2.0/gems/json-1.8.3/tests/test_json.rb:4:in `<top (required)>'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:10:in `require'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:10:in `block (2 levels) in <main>'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:9:in `each'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:9:in `block in <main>'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:4:in `select'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:4:in `<main>'
rake aborted!
Command failed with status (1): [ruby -I"lib"  "/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb" "**/tests/test_*.rb" ]
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/testtask.rb:108:in `block (3 levels) in define'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/file_utils.rb:57:in `call'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/file_utils.rb:57:in `sh'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/file_utils_ext.rb:37:in `sh'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/file_utils.rb:96:in `ruby'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/file_utils_ext.rb:37:in `ruby'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/testtask.rb:104:in `block (2 levels) in define'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/file_utils_ext.rb:58:in `verbose'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/testtask.rb:100:in `block in define'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/task.rb:240:in `call'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/task.rb:240:in `block in execute'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/task.rb:235:in `each'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/task.rb:235:in `execute'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/task.rb:179:in `block in invoke_with_call_chain'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/task.rb:172:in `invoke_with_call_chain'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/task.rb:165:in `invoke'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:150:in `invoke_task'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:106:in `block (2 levels) in top_level'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:106:in `each'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:106:in `block in top_level'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:115:in `run_with_threads'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:100:in `top_level'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:78:in `block in run'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:176:in `standard_exception_handling'
/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/application.rb:75:in `run'
/opt/clusterware/opt/ruby/bin/rake:33:in `<main>'
Tasks: TOP => test
```

Looks like I've messed something up with how things are being loaded but I can't work out what that is, any ideas?

mjtko commented:
> Do I need to update anywhere to include these in the generated config file for new Clusterware installations? These may need documenting for users somewhere also.

Yup, these should be added to the default `gridware.yml` file created as part of `libexec/gridware/init`.

mjtko commented:
> Looks like I've messed something up with how things are being loaded but I can't work out what that is, any ideas?

Hmm. I guess it depends how you've installed the dependencies.  You might need the following?

`/opt/clusterware/opt/ruby/bin/bundle exec rake test`

bobwhitelock commented:
> Hmm. I guess it depends how you've installed the dependencies. You might need the following?
> 
> /opt/clusterware/opt/ruby/bin/bundle exec rake test

This seems to give me the same error:

```
[root@login1(el7cluster) ruby]# /opt/clusterware/opt/ruby/bin/bundle exec rake test
bundler: command not found: rake
Install missing gem executables with `bundle install`
[root@login1(el7cluster) ruby]# /opt/clusterware/opt/ruby/bin/bundle exec /opt/clusterware/opt/ruby/bin/rake test
/opt/clusterware/lib/ruby/vendor/ruby/2.2.0/gems/json-1.8.3/tests/test_json.rb:4:in `require': cannot load such file -- test/unit (LoadError)
        from /opt/clusterware/lib/ruby/vendor/ruby/2.2.0/gems/json-1.8.3/tests/test_json.rb:4:in `<top (required)>'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:10:in `require'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:10:in `block (2 levels) in <main>'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:9:in `each'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:9:in `block in <main>'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:4:in `select'
        from /opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb:4:in `<main>'
rake aborted!
Command failed with status (1): [ruby -I"lib" -I"/opt/clusterware/opt/ruby/lib/ruby/2.2.0" "/opt/clusterware/opt/ruby/lib/ruby/2.2.0/rake/rake_test_loader.rb" "**/tests/test_*.rb" ]

Tasks: TOP => test
(See full trace by running task with --trace)
```

bobwhitelock committed:
Strip trailing whitespace

bobwhitelock committed:
Add initial config values for automatic updates

bobwhitelock commented:
(Just rebased to resolve conflict)

bobwhitelock committed:
Fix test glob to not match files in vendor

Was matching test_* files in vendor/, which was breaking things, so make
glob more specific.
